pep-0592.rst

```diff
@@ -0,0 +1,295 @@
+annotations rules out advanced features only supported by specific type
+checkers.
+
+Rationale
+---------
+
+We propose adding an ``Annotated`` type to the typing module to decorate
+existing types with context-specific metadata. Specifically, a type
+``T`` can be annotated with metadata ``x`` via the typehint
+``Annotated[T, x]``. This metadata can be used for either static
+analysis or at runtime. If a library (or tool) encounters a typehint
+``Annotated[T, x]`` and has no special logic for metadata ``x``, it
+should ignore it and simply treat the type as ``T``. Unlike the
+``no_type_check`` functionality that current exists in the ``typing``
+module which completely disables typechecking annotations on a function
+or a class, the ``Annotated`` type allows for both static typechecking
+of ``T`` (e.g., via mypy [mypy]_ or Pyre [pyre]_, which can safely ignore ``x``)
+together with runtime access to ``x`` within a specific application. We
+believe that the introduction of this type would address a diverse set
+of use cases of interest to the broader Python community.
+
+This was originally brought up as issue 600 [issue-600]_ in the typing github
+and then discussed in Python ideas [python-ideas]_.
+
+Motivating examples
+-------------------
+
+Combining runtime and static uses of annotations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+There's an emerging trend of libraries leveraging the typing annotations at
+runtime (e.g.: dataclasses); having the ability to extend the typing annotations
+with external data would be a great boon for those libraries.
+
+Example::
+
+    UnsignedShort = Annotated[int, cstruct.ctype('H')]
+    SignedChar = Annotated[int, cstruct.ctype('b')]
+
+    class Student(cstruct.Packed):
+        # mypy typechecks 'name' field as 'str'
+        name: Annotated[str, cstruct.ctype("<10s")]
+        serialnum: UnsignedShort
+        school: SignedChar
+        gradelevel: SignedChar
+
+    # 'unpack' only uses the metadata within the type annotations
+    Student.unpack(record)
+    # Student(name=b'raymond   ', serialnum=4658, school=264, gradelevel=8)
+
+Lowering barriers to developing new typing constructs
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Typically when adding a new type, we need to upstream that type to the
+typing module and change mypy, PyCharm [pycharm]_, Pyre,
+pytype [pytype]_, etc...
+This is particularly important when working on open-source code that
+makes use of our new types, seeing as the code would not be immediately
+transportable to other developers' tools without additional logic. As a result,
+there is a high cost to developing and trying out new types in a codebase.
+Ideally, we should be able to introduce new types in a manner that allows for
+graceful degradation when clients do not have a custom mypy plugin
+[mypy-plugin]_, which would lower the barrier to development and ensure some
+degree of backward compatibility.
```

Comment:
TBH, this sentence is still not very clear. In particular, it is not clear what is special about mypy here. Maybe add "for example" somewhere?